// @nx/vite/src/utils/generator-utils.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    findExistingTargetsInProject: function() {
        return findExistingTargetsInProject;
    },
    addOrChangeTestTarget: function() {
        return addOrChangeTestTarget;
    },
    addOrChangeBuildTarget: function() {
        return addOrChangeBuildTarget;
    },
    addOrChangeServeTarget: function() {
        return addOrChangeServeTarget;
    },
    addPreviewTarget: function() {
        return addPreviewTarget;
    },
    editTsConfig: function() {
        return editTsConfig;
    },
    deleteWebpackConfig: function() {
        return deleteWebpackConfig;
    },
    moveAndEditIndexHtml: function() {
        return moveAndEditIndexHtml;
    },
    createOrEditViteConfig: function() {
        return createOrEditViteConfig;
    },
    normalizeViteConfigFilePathWithTree: function() {
        return normalizeViteConfigFilePathWithTree;
    },
    getViteConfigPathForProject: function() {
        return getViteConfigPathForProject;
    },
    handleUnsupportedUserProvidedTargets: function() {
        return handleUnsupportedUserProvidedTargets;
    },
    handleUnknownExecutors: function() {
        return handleUnknownExecutors;
    }
});
const _extends = require("@swc/helpers/_/_extends");
const _devkit = require("@nx/devkit");
const _viteconfigeditutils = require("./vite-config-edit-utils");
function findExistingTargetsInProject(targets, userProvidedTargets) {
    const output = {
        validFoundTargetName: {},
        projectContainsUnsupportedExecutor: false,
        userProvidedTargetIsUnsupported: {},
        alreadyHasNxViteTargets: {}
    };
    const supportedExecutors = {
        build: [
            '@nxext/vite:build',
            '@nx/js:babel',
            '@nx/js:swc',
            '@nx/webpack:webpack',
            '@nx/rollup:rollup',
            '@nrwl/js:babel',
            '@nrwl/js:swc',
            '@nrwl/webpack:webpack',
            '@nrwl/rollup:rollup',
            '@nrwl/web:rollup'
        ],
        serve: [
            '@nxext/vite:dev',
            '@nx/webpack:dev-server',
            '@nrwl/webpack:dev-server'
        ],
        test: [
            '@nx/jest:jest',
            '@nrwl/jest:jest',
            '@nxext/vitest:vitest'
        ]
    };
    const unsupportedExecutors = [
        '@nx/angular:ng-packagr-lite',
        '@nx/angular:package',
        '@nx/angular:webpack-browser',
        '@nx/esbuild:esbuild',
        '@nx/react-native:run-ios',
        '@nx/react-native:start',
        '@nx/react-native:run-android',
        '@nx/react-native:bundle',
        '@nx/react-native:build-android',
        '@nx/react-native:bundle',
        '@nx/next:build',
        '@nx/next:server',
        '@nx/js:tsc',
        '@nrwl/angular:ng-packagr-lite',
        '@nrwl/angular:package',
        '@nrwl/angular:webpack-browser',
        '@nrwl/esbuild:esbuild',
        '@nrwl/react-native:run-ios',
        '@nrwl/react-native:start',
        '@nrwl/react-native:run-android',
        '@nrwl/react-native:bundle',
        '@nrwl/react-native:build-android',
        '@nrwl/react-native:bundle',
        '@nrwl/next:build',
        '@nrwl/next:server',
        '@nrwl/js:tsc',
        '@angular-devkit/build-angular:browser',
        '@angular-devkit/build-angular:dev-server'
    ];
    // First, we check if the user has provided a target
    // If they have, we check if the executor the target is using is supported
    // If it's not supported, then we set the unsupported flag to true for that target
    function checkUserProvidedTarget(target) {
        if (userProvidedTargets == null ? void 0 : userProvidedTargets[target]) {
            var _targets_userProvidedTargets_target;
            if (supportedExecutors[target].includes((_targets_userProvidedTargets_target = targets[userProvidedTargets[target]]) == null ? void 0 : _targets_userProvidedTargets_target.executor)) {
                output.validFoundTargetName[target] = userProvidedTargets[target];
            } else {
                output.userProvidedTargetIsUnsupported[target] = true;
            }
        }
    }
    checkUserProvidedTarget('build');
    checkUserProvidedTarget('serve');
    checkUserProvidedTarget('test');
    // Returns early when we have a build, serve, and test targets.
    if (output.validFoundTargetName.build && output.validFoundTargetName.serve && output.validFoundTargetName.test) {
        return output;
    }
    // We try to find the targets that are using the supported executors
    // for build, serve and test, since these are the ones we will be converting
    for(const target in targets){
        var _hasViteTargets, _hasViteTargets1, _hasViteTargets2, _hasViteTargets3, _output;
        const executorName = targets[target].executor;
        const hasViteTargets = output.alreadyHasNxViteTargets;
        (_hasViteTargets = hasViteTargets).build || (_hasViteTargets.build = executorName === '@nx/vite:build' || executorName === '@nrwl/vite:build');
        (_hasViteTargets1 = hasViteTargets).serve || (_hasViteTargets1.serve = executorName === '@nx/vite:dev-server' || executorName === '@nrwl/vite:dev-server');
        (_hasViteTargets2 = hasViteTargets).test || (_hasViteTargets2.test = executorName === '@nx/vite:test' || executorName === '@nrwl/vite:test');
        (_hasViteTargets3 = hasViteTargets).preview || (_hasViteTargets3.preview = executorName === '@nx/vite:preview-server' || executorName === '@nrwl/vite:preview-server');
        const foundTargets = output.validFoundTargetName;
        if (!foundTargets.build && supportedExecutors.build.includes(executorName)) {
            foundTargets.build = target;
        }
        if (!foundTargets.serve && supportedExecutors.serve.includes(executorName)) {
            foundTargets.serve = target;
        }
        if (!foundTargets.test && supportedExecutors.test.includes(executorName)) {
            foundTargets.test = target;
        }
        (_output = output).projectContainsUnsupportedExecutor || (_output.projectContainsUnsupportedExecutor = unsupportedExecutors.includes(executorName));
    }
    return output;
}
function addOrChangeTestTarget(tree, options, target) {
    var _project;
    const project = (0, _devkit.readProjectConfiguration)(tree, options.project);
    const coveragePath = (0, _devkit.joinPathFragments)('coverage', project.root === '.' ? options.project : project.root);
    const testOptions = {
        passWithNoTests: true,
        // vitest runs in the project root so we have to offset to the workspaceRoot
        reportsDirectory: (0, _devkit.joinPathFragments)((0, _devkit.offsetFromRoot)(project.root), coveragePath)
    };
    var _targets;
    (_targets = (_project = project).targets) != null ? _targets : _project.targets = {};
    if (project.targets[target]) {
        var _project_targets_target_options;
        project.targets[target].executor = '@nx/vite:test';
        (_project_targets_target_options = project.targets[target].options) == null ? void 0 : delete _project_targets_target_options.jestConfig;
    } else {
        project.targets[target] = {
            executor: '@nx/vite:test',
            outputs: [
                coveragePath
            ],
            options: testOptions
        };
    }
    (0, _devkit.updateProjectConfiguration)(tree, options.project, project);
}
function addOrChangeBuildTarget(tree, options, target) {
    var _project;
    const project = (0, _devkit.readProjectConfiguration)(tree, options.project);
    const buildOptions = {
        outputPath: (0, _devkit.joinPathFragments)('dist', project.root != '.' ? project.root : options.project)
    };
    var _targets;
    (_targets = (_project = project).targets) != null ? _targets : _project.targets = {};
    if (project.targets[target]) {
        var _project_targets_target_options;
        buildOptions.fileReplacements = (_project_targets_target_options = project.targets[target].options) == null ? void 0 : _project_targets_target_options.fileReplacements;
        if (project.targets[target].executor === '@nxext/vite:build') {
            var _project_targets_target_options1, _project_targets_target_options2;
            buildOptions.base = (_project_targets_target_options1 = project.targets[target].options) == null ? void 0 : _project_targets_target_options1.baseHref;
            buildOptions.sourcemap = (_project_targets_target_options2 = project.targets[target].options) == null ? void 0 : _project_targets_target_options2.sourcemaps;
        }
        project.targets[target].options = _extends._({}, buildOptions);
        project.targets[target].executor = '@nx/vite:build';
    } else {
        project.targets[target] = {
            executor: '@nx/vite:build',
            outputs: [
                '{options.outputPath}'
            ],
            defaultConfiguration: 'production',
            options: buildOptions,
            configurations: {
                development: {
                    mode: 'development'
                },
                production: {
                    mode: 'production'
                }
            }
        };
    }
    (0, _devkit.updateProjectConfiguration)(tree, options.project, project);
}
function addOrChangeServeTarget(tree, options, target) {
    var _project;
    const project = (0, _devkit.readProjectConfiguration)(tree, options.project);
    var _targets;
    (_targets = (_project = project).targets) != null ? _targets : _project.targets = {};
    if (project.targets[target]) {
        var _project_targets_target_options, _project_targets_target_options1, _project_targets_target_options2;
        const serveTarget = project.targets[target];
        const serveOptions = {
            buildTarget: `${options.project}:build`,
            https: (_project_targets_target_options = project.targets[target].options) == null ? void 0 : _project_targets_target_options.https,
            hmr: (_project_targets_target_options1 = project.targets[target].options) == null ? void 0 : _project_targets_target_options1.hmr,
            open: (_project_targets_target_options2 = project.targets[target].options) == null ? void 0 : _project_targets_target_options2.open
        };
        if (serveTarget.executor === '@nxext/vite:dev') {
            serveOptions.proxyConfig = project.targets[target].options.proxyConfig;
        }
        serveTarget.executor = '@nx/vite:dev-server';
        serveTarget.options = serveOptions;
    } else {
        project.targets[target] = {
            executor: '@nx/vite:dev-server',
            defaultConfiguration: 'development',
            options: {
                buildTarget: `${options.project}:build`
            },
            configurations: {
                development: {
                    buildTarget: `${options.project}:build:development`,
                    hmr: true
                },
                production: {
                    buildTarget: `${options.project}:build:production`,
                    hmr: false
                }
            }
        };
    }
    (0, _devkit.updateProjectConfiguration)(tree, options.project, project);
}
function addPreviewTarget(tree, options, serveTarget) {
    var _project;
    const project = (0, _devkit.readProjectConfiguration)(tree, options.project);
    const previewOptions = {
        buildTarget: `${options.project}:build`
    };
    var _targets;
    (_targets = (_project = project).targets) != null ? _targets : _project.targets = {};
    // Update the options from the passed serve target.
    if (project.targets[serveTarget]) {
        var _target_options, _target_options1;
        const target = project.targets[serveTarget];
        if (target.executor === '@nxext/vite:dev') {
            previewOptions.proxyConfig = target.options.proxyConfig;
        }
        previewOptions.https = (_target_options = target.options) == null ? void 0 : _target_options.https;
        previewOptions.open = (_target_options1 = target.options) == null ? void 0 : _target_options1.open;
    }
    // Adds a preview target.
    project.targets.preview = {
        executor: '@nx/vite:preview-server',
        defaultConfiguration: 'development',
        options: previewOptions,
        configurations: {
            development: {
                buildTarget: `${options.project}:build:development`
            },
            production: {
                buildTarget: `${options.project}:build:production`
            }
        }
    };
    (0, _devkit.updateProjectConfiguration)(tree, options.project, project);
}
function editTsConfig(tree, options) {
    const projectConfig = (0, _devkit.readProjectConfiguration)(tree, options.project);
    const config = (0, _devkit.readJson)(tree, `${projectConfig.root}/tsconfig.json`);
    const commonCompilerOptions = {
        target: 'ESNext',
        useDefineForClassFields: true,
        module: 'ESNext',
        strict: true,
        moduleResolution: 'Node',
        resolveJsonModule: true,
        isolatedModules: true,
        types: [
            'vite/client'
        ],
        noEmit: true
    };
    switch(options.uiFramework){
        case 'react':
            config.compilerOptions = _extends._({}, commonCompilerOptions, {
                lib: [
                    'DOM',
                    'DOM.Iterable',
                    'ESNext'
                ],
                allowJs: false,
                esModuleInterop: false,
                skipLibCheck: true,
                allowSyntheticDefaultImports: true,
                forceConsistentCasingInFileNames: true,
                jsx: 'react-jsx'
            });
            config.include = [
                ...config.include,
                'src'
            ];
            break;
        case 'none':
            config.compilerOptions = _extends._({}, commonCompilerOptions, {
                lib: [
                    'ESNext',
                    'DOM'
                ],
                skipLibCheck: true,
                esModuleInterop: true,
                strict: true,
                noUnusedLocals: true,
                noUnusedParameters: true,
                noImplicitReturns: true
            });
            config.include = [
                ...config.include,
                'src'
            ];
            break;
        default:
            break;
    }
    (0, _devkit.writeJson)(tree, `${projectConfig.root}/tsconfig.json`, config);
}
function deleteWebpackConfig(tree, projectRoot, webpackConfigFilePath) {
    const webpackConfigPath = webpackConfigFilePath && tree.exists(webpackConfigFilePath) ? webpackConfigFilePath : tree.exists(`${projectRoot}/webpack.config.js`) ? `${projectRoot}/webpack.config.js` : tree.exists(`${projectRoot}/webpack.config.ts`) ? `${projectRoot}/webpack.config.ts` : null;
    if (webpackConfigPath) {
        tree.delete(webpackConfigPath);
    }
}
function moveAndEditIndexHtml(tree, options, buildTarget) {
    var _projectConfig_targets, _projectConfig_targets_buildTarget, _projectConfig_targets_buildTarget_options, _projectConfig_targets1, _projectConfig_targets_buildTarget1, _projectConfig_targets_buildTarget_options1;
    const projectConfig = (0, _devkit.readProjectConfiguration)(tree, options.project);
    var _projectConfig_targets_buildTarget_options_index;
    let indexHtmlPath = (_projectConfig_targets_buildTarget_options_index = (_projectConfig_targets = projectConfig.targets) == null ? void 0 : (_projectConfig_targets_buildTarget = _projectConfig_targets[buildTarget]) == null ? void 0 : (_projectConfig_targets_buildTarget_options = _projectConfig_targets_buildTarget.options) == null ? void 0 : _projectConfig_targets_buildTarget_options.index) != null ? _projectConfig_targets_buildTarget_options_index : `${projectConfig.root}/src/index.html`;
    var _projectConfig_targets_buildTarget_options_main;
    let mainPath = (_projectConfig_targets_buildTarget_options_main = (_projectConfig_targets1 = projectConfig.targets) == null ? void 0 : (_projectConfig_targets_buildTarget1 = _projectConfig_targets1[buildTarget]) == null ? void 0 : (_projectConfig_targets_buildTarget_options1 = _projectConfig_targets_buildTarget1.options) == null ? void 0 : _projectConfig_targets_buildTarget_options1.main) != null ? _projectConfig_targets_buildTarget_options_main : `${projectConfig.root}/src/main.ts${options.uiFramework === 'react' ? 'x' : ''}`;
    if (projectConfig.root !== '.') {
        mainPath = mainPath.replace(projectConfig.root, '');
    }
    if (!tree.exists(indexHtmlPath) && tree.exists(`${projectConfig.root}/index.html`)) {
        indexHtmlPath = `${projectConfig.root}/index.html`;
    }
    if (tree.exists(indexHtmlPath)) {
        const indexHtmlContent = tree.read(indexHtmlPath, 'utf8');
        if (!indexHtmlContent.includes(`<script type="module" src="${mainPath}"></script>`)) {
            tree.write(`${projectConfig.root}/index.html`, indexHtmlContent.replace('</body>', `<script type="module" src="${mainPath}"></script>
          </body>`));
            if (tree.exists(`${projectConfig.root}/src/index.html`)) {
                tree.delete(`${projectConfig.root}/src/index.html`);
            }
        }
    } else {
        tree.write(`${projectConfig.root}/index.html`, `<!DOCTYPE html>
      <html lang="en">
        <head>
          <meta charset="UTF-8" />
          <link rel="icon" href="/favicon.ico" />
          <meta name="viewport" content="width=device-width, initial-scale=1.0" />
          <title>Vite</title>
        </head>
        <body>
          <div id="root"></div>
          <script type="module" src="${mainPath}"></script>
        </body>
      </html>`);
    }
}
function createOrEditViteConfig(tree, options, onlyVitest, projectAlreadyHasViteTargets) {
    const projectConfig = (0, _devkit.readProjectConfiguration)(tree, options.project);
    const viteConfigPath = `${projectConfig.root}/vite.config.ts`;
    const buildOption = onlyVitest ? '' : options.includeLib ? `
      // Configuration for building your library.
      // See: https://vitejs.dev/guide/build.html#library-mode
      build: {
        lib: {
          // Could also be a dictionary or array of multiple entry points.
          entry: 'src/index.ts',
          name: '${options.project}',
          fileName: 'index',
          // Change this to the formats you want to support.
          // Don't forgot to update your package.json as well.
          formats: ['es', 'cjs']
        },
        rollupOptions: {
          // External packages that should not be bundled into your library.
          external: [${options.uiFramework === 'react' ? "'react', 'react-dom', 'react/jsx-runtime'" : ''}]
        }
      },` : ``;
    const dtsPlugin = onlyVitest ? '' : options.includeLib ? `dts({
      entryRoot: 'src',
      tsConfigFilePath: joinPathFragments(__dirname, 'tsconfig.lib.json'),
      skipDiagnostics: true,
    }),` : '';
    const dtsImportLine = onlyVitest ? '' : options.includeLib ? `import dts from 'vite-plugin-dts';\nimport { joinPathFragments } from '@nx/devkit';` : '';
    let viteConfigContent = '';
    const testOption = options.includeVitest ? `test: {
    globals: true,
    cache: {
      dir: '${(0, _devkit.offsetFromRoot)(projectConfig.root)}node_modules/.vitest'
    },
    environment: 'jsdom',
    include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
    ${options.inSourceTests ? `includeSource: ['src/**/*.{js,mjs,cjs,ts,mts,cts,jsx,tsx}']` : ''}
  },` : '';
    const defineOption = options.inSourceTests ? `define: {
    'import.meta.vitest': undefined
  },` : '';
    const reactPluginImportLine = options.uiFramework === 'react' ? options.compiler === 'swc' ? `import react from '@vitejs/plugin-react-swc';` : `import react from '@vitejs/plugin-react';` : '';
    const reactPlugin = options.uiFramework === 'react' ? `react(),` : '';
    const devServerOption = onlyVitest ? '' : options.includeLib ? '' : `
    server:{
      port: 4200,
      host: 'localhost',
    },`;
    const previewServerOption = onlyVitest ? '' : options.includeLib ? '' : `
    preview:{
      port: 4300,
      host: 'localhost',
    },`;
    const pluginOption = `
    plugins: [
      ${dtsPlugin}
      ${reactPlugin}
      viteTsConfigPaths({
        root: '${(0, _devkit.offsetFromRoot)(projectConfig.root)}',
      }),
    ],
    `;
    const workerOption = `
    // Uncomment this if you are using workers. 
    // worker: {
    //  plugins: [
    //    viteTsConfigPaths({
    //      root: '${(0, _devkit.offsetFromRoot)(projectConfig.root)}',
    //    }),
    //  ],
    // },`;
    const cacheDir = `cacheDir: '${(0, _devkit.offsetFromRoot)(projectConfig.root)}node_modules/.vite/${options.project}',`;
    if (tree.exists(viteConfigPath)) {
        handleViteConfigFileExists(tree, viteConfigPath, options, buildOption, dtsPlugin, dtsImportLine, pluginOption, testOption, cacheDir, (0, _devkit.offsetFromRoot)(projectConfig.root), projectAlreadyHasViteTargets);
        return;
    }
    viteConfigContent = `
      /// <reference types="vitest" />
      import { defineConfig } from 'vite';
      ${reactPluginImportLine}
      import viteTsConfigPaths from 'vite-tsconfig-paths';
      ${dtsImportLine}
      
      export default defineConfig({
        ${cacheDir}
        ${devServerOption}
        ${previewServerOption}
        ${pluginOption}
        ${workerOption}
        ${buildOption}
        ${defineOption}
        ${testOption}
      });`;
    tree.write(viteConfigPath, viteConfigContent);
}
function normalizeViteConfigFilePathWithTree(tree, projectRoot, configFile) {
    return configFile && tree.exists(configFile) ? configFile : tree.exists((0, _devkit.joinPathFragments)(`${projectRoot}/vite.config.ts`)) ? (0, _devkit.joinPathFragments)(`${projectRoot}/vite.config.ts`) : tree.exists((0, _devkit.joinPathFragments)(`${projectRoot}/vite.config.js`)) ? (0, _devkit.joinPathFragments)(`${projectRoot}/vite.config.js`) : undefined;
}
function getViteConfigPathForProject(tree, projectName, target) {
    let viteConfigPath;
    const { targets , root  } = (0, _devkit.readProjectConfiguration)(tree, projectName);
    if (target) {
        var _targets_target, _targets_target_options;
        viteConfigPath = targets == null ? void 0 : (_targets_target = targets[target]) == null ? void 0 : (_targets_target_options = _targets_target.options) == null ? void 0 : _targets_target_options.configFile;
    } else {
        var _config_options;
        const config = Object.values(targets).find((config)=>config.executor === '@nrwl/nx:build' || config.executor === '@nrwl/vite:build');
        viteConfigPath = config == null ? void 0 : (_config_options = config.options) == null ? void 0 : _config_options.configFile;
    }
    return normalizeViteConfigFilePathWithTree(tree, root, viteConfigPath);
}
async function handleUnsupportedUserProvidedTargets(userProvidedTargetIsUnsupported, userProvidedTargetName, validFoundTargetName) {
    if (userProvidedTargetIsUnsupported.build && validFoundTargetName.build) {
        await handleUnsupportedUserProvidedTargetsErrors(userProvidedTargetName.build, validFoundTargetName.build, 'build', 'build');
    }
    if (userProvidedTargetIsUnsupported.serve && validFoundTargetName.serve) {
        await handleUnsupportedUserProvidedTargetsErrors(userProvidedTargetName.serve, validFoundTargetName.serve, 'serve', 'dev-server');
    }
    if (userProvidedTargetIsUnsupported.test && validFoundTargetName.test) {
        await handleUnsupportedUserProvidedTargetsErrors(userProvidedTargetName.test, validFoundTargetName.test, 'test', 'test');
    }
}
async function handleUnsupportedUserProvidedTargetsErrors(userProvidedTargetName, validFoundTargetName, target, executor) {
    _devkit.logger.warn(`The custom ${target} target you provided (${userProvidedTargetName}) cannot be converted to use the @nx/vite:${executor} executor.
     However, we found the following ${target} target in your project that can be converted: ${validFoundTargetName}

     Please note that converting a potentially non-compatible project to use Vite.js may result in unexpected behavior. Always commit
     your changes before converting a project to use Vite.js, and test the converted project thoroughly before deploying it.
    `);
    const { Confirm  } = require('enquirer');
    const prompt = new Confirm({
        name: 'question',
        message: `Should we convert the ${validFoundTargetName} target to use the @nx/vite:${executor} executor?`,
        initial: true
    });
    const shouldConvert = await prompt.run();
    if (!shouldConvert) {
        throw new Error(`The ${target} target ${userProvidedTargetName} cannot be converted to use the @nx/vite:${executor} executor.
      Please try again, either by providing a different ${target} target or by not providing a target at all (Nx will
        convert the first one it finds, most probably this one: ${validFoundTargetName})

      Please note that converting a potentially non-compatible project to use Vite.js may result in unexpected behavior. Always commit
      your changes before converting a project to use Vite.js, and test the converted project thoroughly before deploying it.
      `);
    }
}
async function handleUnknownExecutors(projectName) {
    _devkit.logger.warn(`
      We could not find any targets in project ${projectName} that use executors which 
      can be converted to the @nx/vite executors.

      This either means that your project may not have a target 
      for building, serving, or testing at all, or that your targets are 
      using executors that are not known to Nx.
      
      If you still want to convert your project to use the @nx/vite executors,
      please make sure to commit your changes before running this generator.
      `);
    const { Confirm  } = require('enquirer');
    const prompt = new Confirm({
        name: 'question',
        message: `Should Nx convert your project to use the @nx/vite executors?`,
        initial: true
    });
    const shouldConvert = await prompt.run();
    if (!shouldConvert) {
        throw new Error(`
      Nx could not verify that the executors you are using can be converted to the @nx/vite executors.
      Please try again with a different project.
    `);
    }
}
function handleViteConfigFileExists(tree, viteConfigPath, options, buildOption, dtsPlugin, dtsImportLine, pluginOption, testOption, cacheDir, offsetFromRoot, projectAlreadyHasViteTargets) {
    if (projectAlreadyHasViteTargets.build && projectAlreadyHasViteTargets.test) {
        return;
    }
    _devkit.logger.info(`vite.config.ts already exists for project ${options.project}.`);
    const buildOptionObject = {
        lib: {
            entry: 'src/index.ts',
            name: options.project,
            fileName: 'index',
            formats: [
                'es',
                'cjs'
            ]
        },
        rollupOptions: {
            external: options.uiFramework === 'react' ? [
                'react',
                'react-dom',
                'react/jsx-runtime'
            ] : []
        }
    };
    const testOptionObject = {
        globals: true,
        cache: {
            dir: `${offsetFromRoot}node_modules/.vitest`
        },
        environment: 'jsdom',
        include: [
            'src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'
        ]
    };
    const changed = (0, _viteconfigeditutils.ensureViteConfigIsCorrect)(tree, viteConfigPath, buildOption, buildOptionObject, dtsPlugin, dtsImportLine, pluginOption, testOption, testOptionObject, cacheDir, projectAlreadyHasViteTargets);
    if (!changed) {
        _devkit.logger.warn(`Make sure the following setting exists in your Vite configuration file (${viteConfigPath}):
        
        ${buildOption}
        
        `);
    } else {
        _devkit.logger.info(`
      Vite configuration file (${viteConfigPath}) has been updated with the required settings for the new target(s).
      `);
    }
}
