// @nx/vite/src/executors/dev-server/dev-server.impl.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    viteDevServerExecutor: function() {
        return viteDevServerExecutor;
    },
    default: function() {
        return _default;
    }
});
const _extends = require("@swc/helpers/_/_extends");
require("dotenv/config");
const _vite = require("vite");
const _optionsutils = require("../../utils/options-utils");
const _internal = require("@nx/js/src/internal");
const _path = require("path");
async function* viteDevServerExecutor(options, context) {
    const projectRoot = context.projectsConfigurations.projects[context.projectName].root;
    (0, _internal.registerTsConfigPaths)((0, _path.resolve)(projectRoot, 'tsconfig.json'));
    // Retrieve the option for the configured buildTarget.
    const buildTargetOptions = (0, _optionsutils.getNxTargetOptions)(options.buildTarget, context);
    // Merge the options from the build and dev-serve targets.
    // The latter takes precedence.
    const mergedOptions = _extends._({}, buildTargetOptions, options);
    // Add the server specific configuration.
    const serverConfig = (0, _vite.mergeConfig)((0, _optionsutils.getViteSharedConfig)(mergedOptions, options.clearScreen, context), {
        build: (0, _optionsutils.getViteBuildOptions)(mergedOptions, context),
        server: (0, _optionsutils.getViteServerOptions)(mergedOptions, context)
    });
    try {
        const server = await (0, _vite.createServer)(serverConfig);
        await runViteDevServer(server);
        const resolvedUrls = [
            ...server.resolvedUrls.local,
            ...server.resolvedUrls.network
        ];
        var _resolvedUrls_;
        yield {
            success: true,
            baseUrl: (_resolvedUrls_ = resolvedUrls[0]) != null ? _resolvedUrls_ : ''
        };
    } catch (e) {
        console.error(e);
        yield {
            success: false,
            baseUrl: ''
        };
    }
    await new Promise((resolve)=>{
        process.once('SIGINT', ()=>resolve());
        process.once('SIGTERM', ()=>resolve());
        process.once('exit', ()=>resolve());
    });
}
async function runViteDevServer(server) {
    await server.listen();
    server.printUrls();
    const processOnExit = async ()=>{
        await server.close();
    };
    process.once('SIGINT', processOnExit);
    process.once('SIGTERM', processOnExit);
    process.once('exit', processOnExit);
}
const _default = viteDevServerExecutor;
